File: python/mcap/mcap0/records.py
from dataclasses import dataclass, field
from typing import Dict, List, Tuple

from .data_stream import ReadDataStream, WriteDataStream
from .opcode import Opcode


@dataclass
class McapRecord:
    def write(self, stream: WriteDataStream):
        pass


@dataclass
class Attachment(McapRecord):
    name: str
    created_at: int
    log_time: int
    content_type: str
    data: bytes

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.ATTACHMENT)
        stream.write_prefixed_string(self.name)
        stream.write8(self.created_at)
        stream.write8(self.log_time)
        stream.write_prefixed_string(self.content_type)
        stream.write8(len(self.data))
        stream.write(self.data)
        stream.write4(0)  # TODO: crc
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream):
        name = stream.read_prefixed_string()
        created_at = stream.read8()
        log_time = stream.read8()
        content_type = stream.read_prefixed_string()
        data_length = stream.read8()
        data = stream.read(data_length)
        _crc = stream.read4()
        return Attachment(
            name=name,
            created_at=created_at,
            log_time=log_time,
            content_type=content_type,
            data=data,
        )


@dataclass
class AttachmentIndex(McapRecord):
    offset: int
    length: int
    log_time: int
    data_size: int
    name: str
    content_type: str

    @staticmethod
    def read(stream: ReadDataStream):
        offset = stream.read8()
        length = stream.read8()
        log_time = stream.read8()
        data_size = stream.read8()
        name = stream.read_prefixed_string()
        content_type = stream.read_prefixed_string()
        return AttachmentIndex(
            offset=offset,
            length=length,
            log_time=log_time,
            data_size=data_size,
            name=name,
            content_type=content_type,
        )


@dataclass
class Channel(McapRecord):
    id: int = field(metadata={"value_type": ["int"]})
    topic: str
    message_encoding: str
    metadata: Dict[str, str]
    schema_id: int = field(metadata={"value_type": ["int"]})

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.CHANNEL)
        stream.write2(self.id)
        stream.write2(self.schema_id)
        stream.write_prefixed_string(self.topic)
        stream.write_prefixed_string(self.message_encoding)
        for k, v in self.metadata.items():
            stream.write_prefixed_string(k)
            stream.write_prefixed_string(v)
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream):
        id = stream.read2()
        schema_id = stream.read2()
        topic = stream.read_prefixed_string()
        message_encoding = stream.read_prefixed_string()
        metadata_length = stream.read4()
        metadata_end = stream.count + metadata_length
        metadata: Dict[str, str] = {}
        while stream.count < metadata_end:
            key = stream.read_prefixed_string()
            value = stream.read_prefixed_string()
            metadata[key] = value
        return Channel(
            id=id,
            topic=topic,
            message_encoding=message_encoding,
            metadata=metadata,
            schema_id=schema_id,
        )


@dataclass
class Chunk(McapRecord):
    compression: str
    data: bytes = field(repr=False)
    message_end_time: int
    message_start_time: int
    uncompressed_crc: int
    uncompressed_size: int

    @staticmethod
    def read(stream: ReadDataStream):
        message_start_time = stream.read8()
        message_end_time = stream.read8()
        uncompressed_size = stream.read8()
        uncompressed_crc = stream.read4()
        compression_length = stream.read4()
        compression = str(stream.read(compression_length), "utf-8")
        data_length = stream.read8()
        data = stream.read(data_length)
        return Chunk(
            compression=compression,
            data=data,
            message_end_time=message_end_time,
            message_start_time=message_start_time,
            uncompressed_crc=uncompressed_crc,
            uncompressed_size=uncompressed_size,
        )


@dataclass
class ChunkIndex(McapRecord):
    chunk_length: int
    chunk_start_offset: int
    compression: str
    compressed_size: int
    message_end_time: int
    message_index_length: int
    message_index_offsets: Dict[int, int] = field(
        metadata={"value_type": ["int", "long"]}
    )
    message_start_time: int
    uncompressed_size: int

    @staticmethod
    def read(stream: ReadDataStream):
        message_start_time = stream.read8()
        message_end_time = stream.read8()
        chunk_start_offset = stream.read8()
        chunk_length = stream.read8()
        message_index_offsets_length = stream.read4()
        message_index_offsets: Dict[int, int] = {}
        offsets_end = stream.count + message_index_offsets_length
        while stream.count < offsets_end:
            channel_id = stream.read2()
            channel_offset = stream.read8()
            message_index_offsets[channel_id] = channel_offset
        message_index_length = stream.read8()
        compression = stream.read_prefixed_string()
        compressed_size = stream.read8()
        uncompressed_size = stream.read8()
        return ChunkIndex(
            message_index_offsets=message_index_offsets,
            chunk_start_offset=chunk_start_offset,
            chunk_length=chunk_length,
            compression=compression,
            compressed_size=compressed_size,
            message_end_time=message_end_time,
            message_index_length=message_index_length,
            message_start_time=message_start_time,
            uncompressed_size=uncompressed_size,
        )


@dataclass
class DataEnd(McapRecord):
    data_section_crc: int

    @staticmethod
    def read(stream: ReadDataStream):
        data_section_crc = stream.read4()
        return DataEnd(data_section_crc=data_section_crc)


@dataclass
class Footer(McapRecord):
    summary_start: int
    summary_offset_start: int
    summary_crc: int = field(metadata={"value_type": ["int"]})

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.FOOTER)
        stream.write8(self.summary_start)
        stream.write8(self.summary_offset_start)
        stream.write4(self.summary_crc)
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream):
        summary_start = stream.read8()
        summary_offset_start = stream.read8()
        summary_crc = stream.read4()
        return Footer(
            summary_start=summary_start,
            summary_offset_start=summary_offset_start,
            summary_crc=summary_crc,
        )


@dataclass
class Header(McapRecord):
    profile: str
    library: str

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.HEADER)
        stream.write_prefixed_string(self.profile)
        stream.write_prefixed_string(self.library)
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream):
        profile = stream.read_prefixed_string()
        library = stream.read_prefixed_string()
        return Header(profile, library)


@dataclass
class Message(McapRecord):
    channel_id: int = field(metadata={"value_type": ["int"]})
    log_time: int
    data: bytes
    publish_time: int
    sequence: int = field(metadata={"value_type": ["int"]})

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.MESSAGE)
        stream.write2(self.channel_id)
        stream.write4(self.sequence)
        stream.write8(self.log_time)
        stream.write8(self.publish_time)
        stream.write(self.data)
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream, length: int):
        channel_id = stream.read2()
        sequence = stream.read4()
        log_time = stream.read8()
        publish_time = stream.read8()
        data = stream.read(length - 22)
        return Message(
            channel_id=channel_id,
            log_time=log_time,
            data=data,
            publish_time=publish_time,
            sequence=sequence,
        )


@dataclass
class MessageIndex(McapRecord):
    channel_id: int = field(metadata={"value_type": ["int"]})
    records: List[Tuple[int, int]] = field(metadata={"value_type": ["long", "long"]})

    @staticmethod
    def read(stream: ReadDataStream):
        channel_id = stream.read2()
        records_length = stream.read4()
        entries: list[Tuple[int, int]] = []
        records_end = stream.count + records_length
        while stream.count < records_end:
            timestamp = stream.read8()
            offset = stream.read8()
            entries.append((timestamp, offset))
        return MessageIndex(channel_id, entries)


@dataclass
class MetadataIndex(McapRecord):
    offset: int
    length: int
    name: str

    @staticmethod
    def read(stream: ReadDataStream):
        offset = stream.read8()
        length = stream.read8()
        name = stream.read_prefixed_string()
        return MetadataIndex(offset=offset, length=length, name=name)


@dataclass
class Schema(McapRecord):
    id: int = field(metadata={"value_type": ["int"]})
    data: bytes
    encoding: str
    name: str

    def write(self, stream: WriteDataStream):
        stream.start_record(Opcode.SCHEMA)
        stream.write2(self.id)
        stream.write_prefixed_string(self.name)
        stream.write_prefixed_string(self.encoding)
        stream.write4(len(self.data))
        stream.write(self.data)
        stream.finish_record()

    @staticmethod
    def read(stream: ReadDataStream):
        id = stream.read2()
        name = stream.read_prefixed_string()
        encoding = stream.read_prefixed_string()
        data_length = stream.read4()
        data = stream.read(data_length)
        return Schema(id=id, name=name, encoding=encoding, data=data)


@dataclass
class Statistics(McapRecord):
    attachment_count: int = field(metadata={"value_type": ["int"]})
    channel_count: int = field(metadata={"value_type": ["int"]})
    channel_message_counts: Dict[int, int] = field(
        metadata={"value_type": ["int", "long"]}
    )
    chunk_count: int = field(metadata={"value_type": ["int"]})
    message_count: int
    metadata_count: int = field(metadata={"value_type": ["int"]})
    schema_count: int = field(metadata={"value_type": ["int"]})

    @staticmethod
    def read(stream: ReadDataStream):
        message_count = stream.read8()
        schema_count = stream.read2()
        channel_count = stream.read4()
        attachment_count = stream.read4()
        metadata_count = stream.read4()
        chunk_count = stream.read4()
        channel_message_counts_length = stream.read4()
        message_counts: Dict[int, int] = {}
        counts_end = stream.count + channel_message_counts_length
        while stream.count < counts_end:
            channel_id = stream.read2()
            message_count = stream.read8()
            message_counts[channel_id] = message_count
        return Statistics(
            attachment_count=attachment_count,
            channel_count=channel_count,
            channel_message_counts=message_counts,
            chunk_count=chunk_count,
            message_count=message_count,
            metadata_count=metadata_count,
            schema_count=schema_count,
        )


@dataclass
class SummaryOffset(McapRecord):
    group_opcode: int = field(metadata={"value_type": ["int"]})
    group_start: int
    group_length: int

    @staticmethod
    def read(stream: ReadDataStream):
        group_opcode = stream.read1()
        group_start = stream.read8()
        group_length = stream.read8()
        return SummaryOffset(
            group_opcode=group_opcode,
            group_start=group_start,
            group_length=group_length,
        )

File: swift/mcap/Sources/mcap/Records.swift
import struct Foundation.Data

public typealias SchemaID = UInt16
public typealias ChannelID = UInt32
public typealias Timestamp = UInt64

public enum Opcode: UInt8 {
  case Header = 0x01
  case Footer = 0x02
  case Schema = 0x03
  case Channel = 0x04
  case Message = 0x05
  case Chunk = 0x06
  case MessageIndex = 0x07
  case ChunkIndex = 0x08
  case Attachment = 0x09
  case AttachmentIndex = 0x0A
  case Statistics = 0x0B
  case Metadata = 0x0C
  case MetadataIndex = 0x0D
  case SummaryOffset = 0x0E
  case DataEnd = 0x0F
}

public protocol Record {
  static var opcode: Opcode { get }
}

func prefixedStringLength(_ str: String) -> Int {
  return MemoryLayout<UInt32>.size + str.utf8.count
}

extension Data {
  mutating func append<T>(unsafeBytesOf value: T) {
    Swift.withUnsafeBytes(of: value) {
      append($0.bindMemory(to: UInt8.self))
    }
  }
  
  mutating func appendPrefixedString(_ str: String) {
    var str = str  // withUTF8 may mutate str
    str.withUTF8 {
      append(unsafeBytesOf: UInt32($0.count).littleEndian)
      append($0)
    }
  }
}

public struct Header {
  static let opcode = Opcode.Header
  let profile: String
  let library: String
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + prefixedStringLength(profile) + prefixedStringLength(library))
    data.append(Self.opcode.rawValue)
    data.appendPrefixedString(profile)
    data.appendPrefixedString(library)
  }
}

public struct Footer {
  static let opcode = Opcode.Footer
  let summaryStart: UInt64
  let summaryOffsetStart: UInt64
  let summaryCRC: UInt32
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt32>.size)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: summaryStart.littleEndian)
    data.append(unsafeBytesOf: summaryOffsetStart.littleEndian)
    data.append(unsafeBytesOf: summaryCRC.littleEndian)
  }
}

public struct Schema {
  static let opcode = Opcode.Schema
  let id: SchemaID
  let name: String
  let encoding: String
  let data: Data
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<SchemaID>.size + prefixedStringLength(name) + prefixedStringLength(encoding) + self.data.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: id.littleEndian)
    data.appendPrefixedString(name)
    data.appendPrefixedString(encoding)
    data.append(self.data)
  }
}

public struct Channel {
  static let opcode = Opcode.Channel
  let id: ChannelID
  let schemaID: SchemaID
  let topic: String
  let messageEncoding: String
  let metadata: [String: String]
  
  //unchecked
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<ChannelID>.size + MemoryLayout<SchemaID>.size + prefixedStringLength(topic) + prefixedStringLength(messageEncoding) + MemoryLayout<UInt32>.size * metadata.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: id.littleEndian)
    data.append(unsafeBytesOf: schemaID.littleEndian)
    data.appendPrefixedString(topic)
    data.appendPrefixedString(messageEncoding)
    data.append(unsafeBytesOf: UInt32(metadata.count).littleEndian)
    for (key, value) in metadata {
      data.appendPrefixedString(key)
      data.appendPrefixedString(value)
    }
  }
}

public struct Message {
  static let opcode = Opcode.Message
  let channelID: ChannelID
  let sequence: UInt32
  let logTime: Timestamp
  let publishTime: Timestamp
  let data: Data
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<ChannelID>.size + MemoryLayout<UInt32>.size + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + self.data.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: channelID.littleEndian)
    data.append(unsafeBytesOf: sequence.littleEndian)
    data.append(unsafeBytesOf: logTime.littleEndian)
    data.append(unsafeBytesOf: publishTime.littleEndian)
    data.append(self.data)
  }
}

public struct Chunk {
  static let opcode = Opcode.Chunk
  let messageStartTime: Timestamp
  let messageEndTime: Timestamp
  let uncompressedSize: UInt64
  let uncompressedCRC: UInt32
  let compression: String
  let records: Data
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt32>.size + prefixedStringLength(compression) + self.records.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: messageStartTime.littleEndian)
    data.append(unsafeBytesOf: messageEndTime.littleEndian)
    data.append(unsafeBytesOf: uncompressedSize.littleEndian)
    data.append(unsafeBytesOf: uncompressedCRC.littleEndian)
    data.appendPrefixedString(compression)
    data.append(self.records)
  }
}

public struct MessageIndex {
  static let opcode = Opcode.MessageIndex
  let channelID: ChannelID
  let records: [(logTime: Timestamp, offset: UInt64)]
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<ChannelID>.size + MemoryLayout<UInt32>.size + MemoryLayout<Timestamp>.size + MemoryLayout<UInt64>.size * records.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: channelID.littleEndian)
    data.append(unsafeBytesOf: UInt32(records.count).littleEndian)
    for (logTime, offset) in records {
      data.append(unsafeBytesOf: logTime.littleEndian)
      data.append(unsafeBytesOf: offset.littleEndian)
    }
  }
}

public struct ChunkIndex {
  static let opcode = Opcode.ChunkIndex
  let messageStartTime: Timestamp
  let messageEndTime: Timestamp
  let chunkStartOffset: UInt64
  let chunkLength: UInt64
  let messageIndexOffsets: [ChannelID: UInt64]
  let messageIndexLength: UInt64
  let compression: String
  let compressedSize: UInt64
  let uncompressedSize: UInt64
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + prefixedStringLength(compression))
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: messageStartTime.littleEndian)
    data.append(unsafeBytesOf: messageEndTime.littleEndian)
    data.append(unsafeBytesOf: chunkStartOffset.littleEndian)
    data.append(unsafeBytesOf: chunkLength.littleEndian)
    data.append(unsafeBytesOf: messageIndexLength.littleEndian)
    data.append(unsafeBytesOf: compressedSize.littleEndian)
    data.append(unsafeBytesOf: uncompressedSize.littleEndian)
    data.appendPrefixedString(compression)
    for (channelID, offset) in messageIndexOffsets {
      data.append(unsafeBytesOf: channelID.littleEndian)
      data.append(unsafeBytesOf: offset.littleEndian)
    }
  }
}

public struct Attachment {
  static let opcode = Opcode.Attachment
  let logTime: Timestamp
  let createTime: Timestamp
  let name: String
  let contentType: String
  let data: Data
  @available(*, unavailable, message: "TODO") let crc: UInt32
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + prefixedStringLength(name) + prefixedStringLength(contentType) + self.data.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: logTime.littleEndian)
    data.append(unsafeBytesOf: createTime.littleEndian)
    data.appendPrefixedString(name)
    data.appendPrefixedString(contentType)
    data.append(self.data)
  }
}

public struct AttachmentIndex {
  static let opcode = Opcode.AttachmentIndex
  let offset: UInt64
  let length: UInt64
  let logTime: Timestamp
  let createTime: Timestamp
  let dataSize: UInt64
  let name: String
  let contentType: String
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + MemoryLayout<UInt64>.size + prefixedStringLength(name) + prefixedStringLength(contentType))
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: offset.littleEndian)
    data.append(unsafeBytesOf: length.littleEndian)
    data.append(unsafeBytesOf: logTime.littleEndian)
    data.append(unsafeBytesOf: createTime.littleEndian)
    data.append(unsafeBytesOf: dataSize.littleEndian)
    data.appendPrefixedString(name)
    data.appendPrefixedString(contentType)
  }
}

public struct Statistics {
  static let opcode = Opcode.Statistics
  let messageCount: UInt64
  let schemaCount: UInt16
  let channelCount: UInt32
  let attachmentCount: UInt32
  let metadataCount: UInt32
  let chunkCount: UInt32
  let messageStartTime: Timestamp
  let messageEndTime: Timestamp
  let channelMessageCounts: [ChannelID: UInt64]
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt64>.size + MemoryLayout<UInt16>.size + MemoryLayout<UInt32>.size + MemoryLayout<UInt32>.size + MemoryLayout<UInt32>.size + MemoryLayout<UInt32>.size + MemoryLayout<Timestamp>.size + MemoryLayout<Timestamp>.size + MemoryLayout<UInt32>.size * channelMessageCounts.count)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: messageCount.littleEndian)
    data.append(unsafeBytesOf: schemaCount.littleEndian)
    data.append(unsafeBytesOf: channelCount.littleEndian)
    data.append(unsafeBytesOf: attachmentCount.littleEndian)
    data.append(unsafeBytesOf: metadataCount.littleEndian)
    data.append(unsafeBytesOf: chunkCount.littleEndian)
    data.append(unsafeBytesOf: messageStartTime.littleEndian)
    data.append(unsafeBytesOf: messageEndTime.littleEndian)
    data.append(unsafeBytesOf: UInt32(channelMessageCounts.count).littleEndian)
    for (channelID, count) in channelMessageCounts {
      data.append(unsafeBytesOf: channelID.littleEndian)
      data.append(unsafeBytesOf: count.littleEndian)
    }
  }
}

public struct Metadata {
  static let opcode = Opcode.Metadata
  let name: String
  let metadata: [String: String]
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + prefixedStringLength(name) + MemoryLayout<UInt32>.size + metadata.count * (MemoryLayout<UInt32>.size + prefixedStringLength(key) + prefixedStringLength(value)))
    data.append(Self.opcode.rawValue)
    data.appendPrefixedString(name)
    data.append(UInt32(metadata.count).littleEndian)
    for (key, value) in metadata {
      data.append(UInt32(key.count).littleEndian)
      data.appendPrefixedString(key)
      data.append(UInt32(value.count).littleEndian)
      data.appendPrefixedString(value)
    }
  }
}

public struct MetadataIndex {
  static let opcode = Opcode.MetadataIndex
  let offset: UInt64
  let length: UInt64
  let name: String
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size + prefixedStringLength(name))
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: offset.littleEndian)
    data.append(unsafeBytesOf: length.littleEndian)
    data.appendPrefixedString(name)
  }
}

public struct SummaryOffset {
  static let opcode = Opcode.SummaryOffset
  let groupOpcode: UInt8
  let groupStart: UInt64
  let groupLength: UInt64
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt8>.size + MemoryLayout<UInt64>.size + MemoryLayout<UInt64>.size)
    data.append(Self.opcode.rawValue)
    data.append(groupOpcode)
    data.append(unsafeBytesOf: groupStart.littleEndian)
    data.append(unsafeBytesOf: groupLength.littleEndian)
  }
}

public struct DataEnd {
  static let opcode = Opcode.DataEnd
  let dataSectionCRC: UInt32
  
  func serialize(to data: inout Data) {
    data.reserveCapacity(1 + MemoryLayout<UInt32>.size)
    data.append(Self.opcode.rawValue)
    data.append(unsafeBytesOf: dataSectionCRC.littleEndian)
  }
}

